Client/Application/NavigationRouter.swift

```diff
@@ -0,0 +1,183 @@
+enum HomePanelPath: String {
+    case bookmarks
+    case topsites
+    case readingList
+    case history
+}
+
+// An enum to route to a settings page.
+// This could be extended to provide default values to pass to fxa
+enum SettingsPage: String {
+    case newTab
+    case homePage
+    case mailto
+    case search
+    case clearData = "clear-private-data"
+    case fxa
+}
+
+// Used by the App to navigate to different views.
+// To open a URL use /open-url or to open a blank tab use /open-url with no params
+enum DeepLink {
+    case settings(SettingsPage)
+    case homePanel(HomePanelPath)
+    
+    init?(urlString: String) {
+        if urlString.startsWith("/settings/"), let link = SettingsPage(rawValue: urlString.replacingOccurrences(of: "/settings/", with: "")) {
+            self = .settings(link)
+        } else if urlString.startsWith("/homepanel/"), let link = HomePanelPath(rawValue: urlString.replacingOccurrences(of: "/homepanel/", with: "")) {
+            self = .homePanel(link)
+        } else {
+            return nil
+        }
+    }
+}
+
+extension URLComponents {
+    // Return the first query parameter that matches
+    func valueForQuery(_ param: String) -> String? {
+        return self.queryItems?.first { $0.name == param }?.value
+    }
+}
+    
+// The root navigation for the Router. Look at the tests to see a complete URL
+enum NavigationPath {
+    case url(webURL: URL?, isPrivate: Bool)
+    case fxa(params: FxALaunchParams)
+    case deepLink(DeepLink)
+    
+    init?(url: URL) {
+        let urlString = url.absoluteString
+        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
+            return nil
+        }
+        
+        guard let urlTypes = Bundle.main.object(forInfoDictionaryKey: "CFBundleURLTypes") as? [AnyObject],
+            let urlSchemes = urlTypes.first?["CFBundleURLSchemes"] as? [String] else {
+            // Something very strange has happened; org.mozilla.Client should be the zeroeth URL type.
+            return nil
+        }
+
+        guard let scheme = components.scheme, urlSchemes.contains(scheme) else {
+            return nil
+        }
+        
+        if urlString.startsWith("\(scheme)://deep-link"), let deepURL = components.valueForQuery("url"), let link = DeepLink(urlString: deepURL) {
+            self = .deepLink(link)
+        } else if urlString.startsWith("\(scheme)://fxa-signin"), components.valueForQuery("signin") != nil {
+            self = .fxa(params: FxALaunchParams(query: url.getQuery()))
+        } else if urlString.startsWith("\(scheme)://open-url") {
+            let url = components.valueForQuery("url")?.unescape()?.asURL
+            let isPrivate = Bool(components.valueForQuery("private") ?? "") ?? false
+            self = .url(webURL: url, isPrivate: isPrivate)
+        } else {
+            return nil
+        }
+    }
+    
+    static func handleFxA(params: FxALaunchParams, with bvc: BrowserViewController) {
```

Comment:
Nit: put these static methods in to a `NavigationRouter.shared` stateless singleton.

The entry point should be a `handle(_ url: URL, with bvc: BrowserViewController)`.